compiler/rustc_lint/src/unused.rs

```diff
@@ -88,12 +88,64 @@ declare_lint! {
     "unused result of an expression in a statement"
 }
 
-declare_lint_pass!(UnusedResults => [UNUSED_MUST_USE, UNUSED_RESULTS]);
+declare_lint! {
+    /// The `unused_must_bind` lint detects attempts to "bind" to `_` value of a type marked with
+    /// `#[must_bind]`.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(unused_must_bind)]
+    ///
+    /// #[must_use]
+    /// #[must_bind]
+    /// struct SemaphoreGuard;
+    ///
+    /// fn acquire() -> SemaphoreGuard {
+    ///     SemaphoreGuard
+    /// }
+    ///
+    /// fn main() {
+    ///     let _ = acquire();
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// The `#[must_bind]` attribute is an indicator that it is a mistake to
+    /// let the value be dropped imediately after acquiring it. See [the reference] for more details.
+    ///
+    /// [the reference]: https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_bind-attribute
+    pub UNUSED_MUST_BIND,
+    Warn,
+    "Bogus `_` binding of a type flagged as `#[must_bind]`. Use something like `_guard`.",
+    report_in_external_macro
+}
+
+declare_lint_pass!(UnusedResults => [UNUSED_MUST_USE, UNUSED_RESULTS, UNUSED_MUST_BIND]);
 
 impl<'tcx> LateLintPass<'tcx> for UnusedResults {
     fn check_stmt(&mut self, cx: &LateContext<'_>, s: &hir::Stmt<'_>) {
-        let expr = match s.kind {
-            hir::StmtKind::Semi(ref expr) => &**expr,
+        #[derive(Clone, Copy)]
+        enum WhatAttrToCheck {
+            MustUse,
+            MustBind,
+        }
+        let (expr, attr_to_check) = match s.kind {
+            hir::StmtKind::Semi(ref expr) => (&**expr, WhatAttrToCheck::MustUse),
+            hir::StmtKind::Local(ref local) => {
+                match local.pat.kind {
```

Comment:
I'd prefer a guard in the match arm: `if !matches!(local.pat.kind, hir::PatKind::Wild)`